swift/example_code/s3/basics/Tests/basics-tests/basics-tests.swift

```diff
@@ -0,0 +1,429 @@
+        do {
+            let srcBucketName = try await createTestBucket()
+            let destBucketName = try await createTestBucket()
+        
+            // Create a file.
+
+            let srcFileInfo = try await createTestFile(bucket: srcBucketName)
+
+            // Copy the file.
+
+            let destFileInfo = try await copyTestFile(fileInfo: srcFileInfo, to: destBucketName)
+            if try await verifyTestFileContents(fileInfo: destFileInfo) == false {
+                XCTFail("Copied file does not match the original")
+            }
+        } catch {
+            throw error
+        }
+    }
+
+    /// Test that trying to copy a file into a bucket that doesn't exist fails
+    /// as expected.
+    func testCopyFileInvalidDestination() async throws {
+        do {
+            let srcBucketName = try await createTestBucket()
+            let destBucketName = String.uniqueName()
+
+            // Create the test file.
+
+            let srcFileInfo = try await createTestFile(bucket: srcBucketName)
+
+            do {
+                // Attempt to copy the file.
+
+                let destFileInfo = try await copyTestFile(fileInfo: srcFileInfo, to: destBucketName)
+                XCTFail("Copying file to a non-existent bucket didn't fail like it should")
+            } catch {
+                return      // An error is a success case here
+            }
+        } catch {
+            throw error
+        }
+    }
+
+    /// Test listing files using the `ServiceHandler` function
+    /// `listBucketFiles(bucket:)`.
+    func testListFiles() async throws {
+        do {
+            let bucketName = try await createTestBucket()
+
+            // Create a bunch of files in the bucket, with random names and
+            // contents.
+            for _ in 0...14 {
+                let fileInfo = try await createTestFile(bucket: bucketName,
+                                withParagraphs: Int.random(in: 1...15))
+            }
+
+            // Get a list of the contents of the bucket
+
+            let fileNames = try await listTestFiles(bucket: bucketName)
+            let listedFileCount = fileNames.count
+
+            XCTAssertEqual(15, listedFileCount, "Should be 15 files in the list but actual number is \(listedFileCount)")
+
+            // Build a list of the names we expect to see.
+
+            var expectedNames: [String] = []
+
+            for file in BasicsTests.demoCleanup!.files {
+                if file.bucket == bucketName {
+                    expectedNames.append(file.file)
+                }
+            }
+
+            let expectedFileCount = expectedNames.count
+
+            XCTAssertEqual(listedFileCount, expectedFileCount, "Number of files found in the bucket (\(listedFileCount)) doesn't equal the expected number (\(expectedFileCount))")
+
+            // Make sure that both every file that exists is one we expect, and
+            // that every file we expect to see is present.
```

Comment:
// Make sure that every file that exists is one we expect, and
            // also make sure that every file we expect to see is present.